server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/PartitionCache.java

```diff
@@ -0,0 +1,710 @@
+  public String getStorageGroup(String devicePath) {
+    synchronized (storageGroupCache) {
+      for (String storageGroup : storageGroupCache) {
+        if (PathUtils.isStartWith(devicePath, storageGroup)) {
+          return storageGroup;
+        }
+      }
+    }
+    return null;
+  }
+
+  /** get device to storage group map */
+  public boolean getDeviceToStorageGroupMap(
+      List<String> devicePaths,
+      Map<String, String> deviceToStorageGroupMap,
+      List<String> failedDevices,
+      boolean failFast) {
+    storageGroupCacheLock.readLock().lock();
+    try {
+      boolean result = true;
+      for (String devicePath : devicePaths) {
+        String storageGroup = getStorageGroup(devicePath);
+        if (null == storageGroup) {
+          logger.debug(
+              "[{} Cache] miss when search device {}", STORAGE_GROUP_CACHE_NAME, devicePath);
+          result = false;
+          if (failFast) {
+            break;
+          } else {
+            failedDevices.add(devicePath);
+          }
+        } else {
+          deviceToStorageGroupMap.put(devicePath, storageGroup);
+        }
+      }
+      if (!result) {
+        deviceToStorageGroupMap.clear();
+      }
+      logger.debug("[{} Cache] hit when search device {}", STORAGE_GROUP_CACHE_NAME, devicePaths);
+      CacheMetricsRecorder.record(result, STORAGE_GROUP_CACHE_NAME);
+      return result;
+    } finally {
+      storageGroupCacheLock.readLock().unlock();
+    }
+  }
+
+  /** get storage group to device map with failed devices */
+  public boolean getStorageGroupToDeviceMap(
+      List<String> devicePaths,
+      Map<String, List<String>> storageGroupToDeviceMap,
+      List<String> failedDevices,
+      boolean failFast) {
+    storageGroupCacheLock.readLock().lock();
+    try {
+      boolean result = true;
+      for (String devicePath : devicePaths) {
+        String storageGroup = getStorageGroup(devicePath);
+        if (null == storageGroup) {
+          logger.debug(
+              "[{} Cache] miss when search device {}", STORAGE_GROUP_CACHE_NAME, devicePath);
+          result = false;
+          if (failFast) {
+            break;
+          } else {
+            failedDevices.add(devicePath);
+          }
+        } else {
+          List<String> devices =
+              storageGroupToDeviceMap.computeIfAbsent(storageGroup, k -> new ArrayList<>());
+          devices.add(devicePath);
+        }
+      }
+      if (!result) {
+        storageGroupToDeviceMap.clear();
+      }
+      logger.debug("[{} Cache] hit when search device {}", STORAGE_GROUP_CACHE_NAME, devicePaths);
+      CacheMetricsRecorder.record(result, STORAGE_GROUP_CACHE_NAME);
+      return result;
+    } finally {
+      storageGroupCacheLock.readLock().unlock();
+    }
+  }
+
+  /** get storage group to device map in three trys */
+  public Map<String, List<String>> getStorageGroupToDevice(
+      List<String> devicePaths, boolean isAutoCreate) {
```

Comment:
I suppose that this interface provides the service of _get StorageGroup to Device_. We need to extract the _getStorageGroupToDeviceMap_ thus we can facilitate the retry logic inside this interface. Therefore, I wonder that would it be better if we make the _getStorageGroupToDeviceMap_ **private**?

By the way, we need a more detailed function annotation, which including the definition of every corresponding parameters, if my hypothesize is correct~

Reply:
Get it ~